lib/matplotlib/figure.py

```diff
@@ -1461,8 +1461,41 @@ def legend(self, *args, **kwargs):
         :ref:`sphx_glr_tutorials_intermediate_legend_guide.py` for details.
         """
 
+        # Support handles and labels being passed as keywords.
+        handles = kwargs.pop('handles', None)
+        if handles and not iterable(handles):
+            handles = [handles]
+        labels = kwargs.pop('labels', None)
+        if labels and not iterable(labels):
+            labels = [labels]
+
+        if (handles is not None
+                or labels is not None) and len(args):
+            warnings.warn("You have mixed positional and keyword "
+                          "arguments to legend, some input may be "
+                          "discarded.")
+
+        # if there is just one arg then it is the handles we want to use
+        if not handles and (len(args) == 1):
+            handles = args[0]
+        # if got both handles and labels as kwargs, make same length
+        if handles and labels:
+            handles, labels = zip(*zip(handles, labels))
+
+        # if handles provided but not labels, then they must be
+        # provided in the axes
+        elif handles is not None and labels is None:
+            labels = [handle.get_label() for handle in handles]
+            for label, handle in zip(labels[:], handles[:]):
+                if label.startswith('_'):
+                    warnings.warn('The handle {!r} has a label of {!r} which '
+                                  'cannot be automatically added to the '
+                                  'legend.'.format(handle, label))
+                    labels.remove(label)
```

Comment:
Can we let this be handled down stream?

Reply:
The underscore check?  Maybe.  I just shamelessly copied from axes.legend. 

Yes yes I know I should have refactored it... otoh I don’t like that axes.legend if there is only one argument assumes that it is a list of labels.  I think a list of handles makes much more sense.

Reply:
Put this check into `legend.Legend.__init__`, and removed from `axes.legend` and `figure.legend`

Reply:
Taking labels (not handles) is a bit problematic and leads to issues like https://github.com/benlaken/European_wind/pull/1

Reply:
@tacaswell. Um yeah that does look to be a bad decision. 

I’d suggest we just strip the illegal underscore and warn (or logging.warning 😉) or set it to an empty string.  

But I didn’t originate this code.  I’m not actually sure why an underscore is a problem in the first place.